france_connect/clients.py

```diff
@@ -0,0 +1,192 @@
+        self.verify_ssl = verify_ssl
+        self.allow_redirects = allow_redirects
+
+    def generate_nonce(self) -> str:
+        """Generate a random nonce in hexadecimal string (64 characters).."""
+        return secrets.token_hex(48)
+
+    def generate_state(self) -> str:
+        """Generate a random state in hexadecimal string (64 characters)."""
+        return secrets.token_hex(48)
+
+    def get_authentication_url(
+        self, acr_values: List[ACRValues] = None, nonce: str = None, state: str = None
+    ) -> Tuple[str, str, str]:
+        """Return the authentication URL, nonce and state."""
+        acr_values = acr_values or [ACRValues.EIDAS1]
+        nonce = nonce or self.generate_nonce()
+        state = state or self.generate_state()
+        params = urllib.parse.urlencode(
+            {
+                "client_id": self.client_id,
+                "response_type": "code",
+                "scope": " ".join(self.scopes),
+                "acr_values": " ".join(acr_values),
+                "nonce": nonce,
+                "state": state,
+                "redirect_uri": self.post_login_callback_url,
+            }
+        )
+        url = urllib.parse.urljoin(self.fc_base_url, self.fc_authorization_url)
+        return f"{url}?{params}", nonce, state
+
+    def get_configuration(self) -> dict:
+        """Return the OpenID configuration."""
+        url = urllib.parse.urljoin(self.fc_base_url, "/api/v2/.well-known/openid-configuration")
+        try:
+            response = requests.get(
+                url, timeout=self.timeout, verify=self.verify_ssl, allow_redirects=self.allow_redirects
+            )
+            return response.json()
+        except requests.exceptions.RequestException:
+            msg = f"FranceConnect - Failed to retrieve configuration"
+            raise FranceConnectConfigurationError(response.status_code, response.text, msg)
+
+    def get_access_token(self, code: str) -> dict:
+        """Exchange the authorization code for an access token."""
+        token_url = urllib.parse.urljoin(self.fc_base_url, self.fc_token_url)
+        data = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self.post_login_callback_url,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+        try:
+            response = requests.post(
+                token_url,
+                data=data,
+                headers={"Accept": "application/json"},
+                timeout=self.timeout,
+                verify=self.verify_ssl,
+                allow_redirects=self.allow_redirects,
+            )
+            response.raise_for_status()
+            return response.json()
+        except requests.exceptions.RequestException:
+            msg = f"FranceConnect - Failed to retrieve access_token"
+            raise FranceConnectAccessTokenError(response.status_code, response.text, msg)
```

Comment:
I didn't know the `response.raise_for_status()` method, it's super useful.

You should not handle any error yourself and let `requests` raise its exceptions. it contains everything the package's user needs to handle the error, i.e the following is good enough :

```
    def get_access_token(self, code: str) -> dict:
        """Exchange the authorization code for an access token."""
        token_url = urllib.parse.urljoin(self.fc_base_url, self.fc_token_url)
        data = {
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": self.post_login_callback_url,
            "client_id": self.client_id,
            "client_secret": self.client_secret,
        }
        response.raise_for_status()
        return response.json()
```

You probably can do the same everywhere we raise custom exceptions since none of them add useful information atop the exception raised by the package you're using.

Sorry for misleading you earlier about custom exceptions.